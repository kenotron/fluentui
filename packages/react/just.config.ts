import { task, series, nodeExecTask, webpackCliTask, preset, resolveCwd } from '@fluentui/scripts';

preset();

task(
  'mf',
  series(webpackCliTask({ webpackCliArgs: ['--config', 'webpack.mf.config.js', '--progress'] }), () => {
    const httpServer = resolveCwd('http-server/bin/http-server');
    return nodeExecTask({ args: [httpServer, '-p', '2345', './dist'] });
  }),
);
