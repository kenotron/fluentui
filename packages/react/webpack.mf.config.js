// @ts-check
const resources = require('../../scripts/webpack/webpack-resources');
const webpack = require('webpack');
const { webpackMerge } = require('just-scripts');

const BUNDLE_NAME = 'fluentui-react';

let config = resources.createConfig(BUNDLE_NAME, false, {
  resolve: {
    extensions: ['.ts', '.tsx', '.js'],
  },

  module: {
    rules: [
      {
        test: /\.css$/,
        include: /node_modules/,
        use: ['style-loader', 'css-loader'],
      },
      {
        test: [/\.tsx?$/],
        use: {
          loader: 'ts-loader',
          options: {
            experimentalWatchApi: true,
            transpileOnly: true,
          },
        },
        exclude: [/node_modules/, /\.scss.ts$/, /\.test.tsx?$/],
      },
      {
        test: /\.scss$/,
        enforce: 'pre',
        exclude: [/node_modules/],
        use: [
          {
            loader: '@microsoft/loader-load-themed-styles', // creates style nodes from JS strings
          },
          {
            loader: 'css-loader', // translates CSS into CommonJS
            options: {
              esModule: false,
              modules: true,
              importLoaders: 2,
            },
          },
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: ['autoprefixer'],
              },
            },
          },
          {
            loader: 'sass-loader',
          },
        ],
      },
    ],
  },

  // @ts-ignore
  plugins: [
    new webpack.container.ModuleFederationPlugin({
      name: 'fluentuiReact',
      filename: 'remoteEntry.js',
      exposes: {
        './Button': './src/Button',
      },
    }),
  ],
  output: {
    filename: `${BUNDLE_NAME}.mf.js`,
  },
})[0];

delete config.entry;

module.exports = config;
