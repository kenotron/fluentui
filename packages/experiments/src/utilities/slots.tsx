import * as React from 'react';
// TODO: pull this from utilities instead of adding a dependency to OUFR in Foundation
import { memoizeFunction } from 'office-ui-fabric-react';

/**
 * Signature of components that have component factories.
 */
export type IFactoryComponent<TProps> = React.ReactType<TProps> & {
  create?: ISlotFactory<TProps>;
};

/**
 * Factory options for creating component.
 */
export interface IFactoryOptions<TProps> {
  /** Default prop for which to map primitive values. */
  defaultProp: keyof TProps | 'children';
}

/**
 * Helper interface for accessing user props children.
 */
export type IPropsWithChildren<TProps> = TProps & { children?: React.ReactNode };

/**
 * An interface for defining Slots. Each key in TSlot must point to an IFactoryComponent.
 */
export type ISlotDefinition<TSlots> = { [prop in keyof TSlots]: IFactoryComponent<TSlots[prop]> };

/**
 * Created Slot structure used for rendering by components.
 */
export type ISlot<TProps> = ((props: IPropsWithChildren<TProps>) => JSX.Element) & { isSlot?: boolean };

/**
 * Interface for a slot factory that consumes both componnent and user slot prop and generates rendered output.
 */
export type ISlotFactory<TProps> = (componentProps: TProps, userProps?: ISlotProp<TProps>) => JSX.Element;

/**
 * Interface for aggregated Slots objects used internally by components.
 */
export type ISlots<TSlots> = { [slot in keyof TSlots]: ISlot<TSlots[slot]> };

/**
 * User properties that are automatically applied by Slot utilities using slot name.
 */
export interface IUserProps<TSlots> {
  classNames: { [prop in keyof TSlots]?: string };
}

/**
 * Helper interface components can use for defining Slot properties. This interface defines the following slot properties:
 *    1. Component props object (defined by TProps.)
 *    2. ISlotRender function.
 *    3. JSX Elements.
 *    4. Optional shorthand prop, defined by TShorthandProp.
 * The conditional type check looks up prop type in TProps if TShorthandProp is a key of TProps, otherwise it treats
 * TShorthandProp as React children. If TShorthandProp is excluded, there is no default prop and no children are allowed.
 */
export type ISlotProp<TProps, TShorthandProp extends keyof TProps | 'children' = never> =
  | TProps
  | JSX.Element
  | ISlotRenderFunction<TProps>
  | (TShorthandProp extends keyof TProps ? TProps[TShorthandProp] : React.ReactNode);

/**
 * Render function interface used by Slot props.
 */
export type ISlotRenderFunction<TProps> = (props: TProps, componentType: React.ReactType<TProps>) => JSX.Element;

/**
 * This function removes Slots from the React hierarchy by wrapping React.createElement and bypassing it for Slot components.
 *
 * To use this function on a per-file basis, put the following in a comment block: @jsx createElementWrapper
 * As of writing, this line must be the FIRST LINE in the file to work correctly.
 *
 * Usage of this pragma also requires an import statement of SlotModule such as: import { createElementWrapper } from '<path>/slots';
 *
 * @see React.createElement
 */
// Can't use typeof on React.createElement since it's overloaded. Approximate createElement's signature for now and widen as needed.
export function createElementWrapper<P>(
  type: ISlot<P> | React.SFC<P> | string,
  props?: React.Attributes & P | null,
  // tslint:disable-next-line:missing-optional-annotation
  ...children: React.ReactNode[]
): React.ReactElement<P> | JSX.Element | null {
  const slotType = type as ISlot<P>;
  if (slotType.isSlot) {
    // Since we are bypassing createElement, use React.Children.toArray to make sure children are properly assigned keys.
    children = React.Children.toArray(children);
    return slotType({ ...(props as any), children });
  } else {
    return React.createElement(type, props, children);
  }
}

/**
 * This function creates factories that render ouput depending on the user ISlotProp props passed in.
 * @param Component Base component to render when not overridden by user props.
 * @param options Factory options, including defaultProp value for shorthand prop mapping.
 * @returns ISlotFactory function used for rendering slots.
 */
export function createFactory<TProps>(
  Component: React.ComponentType<TProps>,
  options: IFactoryOptions<TProps> = { defaultProp: 'children' }
): ISlotFactory<TProps> {
  const result: ISlotFactory<TProps> = (componentProps, userProps) => {
    if (userProps) {
      const propType = typeof userProps;

      switch (propType) {
        case 'string':
        case 'number':
        case 'boolean':
          userProps = {
            [options.defaultProp]: userProps as any
          } as TProps;
          break;

        case 'function':
          // Functional components are not identified as functions because they have been converted to React Elements before this point.
          return (userProps as ISlotRenderFunction<TProps>)(componentProps, Component);

        default:
          if (React.isValidElement(userProps)) {
            return userProps;
          }
          break;
      }

      return <Component {...componentProps} {...userProps} />;
    } else {
      return <Component {...componentProps} />;
    }
  };

  return result;
}

/**
 * Default factory for components without explicit factories.
 */
const getDefaultFactory = memoizeFunction(type => createFactory(type));

/**
 * Default factory helper.
 */
function defaultFactory<TComponent, TProps>(type: TComponent, componentProps: TProps, userProps: TProps) {
  return getDefaultFactory(type)(componentProps, userProps);
}

/**
 * Render a slot given component and user props. Uses component factory if available, otherwise falls back
 * to default factory.
 * @param ComponentType Factory component type.
 * @param componentProps The properties passed into slot from within the component.
 * @param userProps The user properties passed in from outside of the component.
 */
function renderSlot<TComponent extends IFactoryComponent<TProps>, TProps>(
  ComponentType: TComponent,
  componentProps: TProps,
  userProps: TProps
): JSX.Element {
  if (ComponentType.create !== undefined) {
    return ComponentType.create(componentProps, userProps);
  } else {
    return defaultFactory(ComponentType, componentProps, userProps);
  }
}

/**
 * This function generates slots that can be used in JSX given a definition of Slots and their corresponding types.
 * @param userProps Props as pass to component.
 * @param slots Slot definition object defining the default slot component for each slot.
 * @returns An set of created slots that components can render in JSX.
 */
export function getSlots<TProps extends TSlots & IUserProps<TSlots>, TSlots extends { [key in keyof TSlots]: ISlotProp<TProps[key]> }>(
  userProps: TProps,
  slots: ISlotDefinition<Required<TSlots>>
): ISlots<Required<TSlots>> {
  const result: ISlots<Required<TSlots>> = {} as ISlots<Required<TSlots>>;

  for (const name in slots) {
    if (slots.hasOwnProperty(name)) {
      if (userProps && userProps[name] && userProps[name].children) {
        // Since createElementWrapper bypasses createElement, use React.Children.toArray to make sure children are properly assigned keys.
        userProps[name].children = React.Children.toArray(userProps[name].children);
      }

      const slot: ISlot<keyof TSlots> = slotProps => {
        return renderSlot(slots[name], { ...(slotProps as any), className: userProps.classNames[name] }, userProps[name]);
      };
      slot.isSlot = true;
      result[name] = slot;
    }
  }

  return result;
}

//////////////////////////////////////////////////////////////////////////////////////////////
// TODO: Slots Phase 2 (future PR)
// TODO: Incorporate style variables approach and lift mergeStyles out of createComponent.
// TODO: Debug scenarios: is there a way to enable or highlight slots visually? borders? etc.
// TODO: Need something like this to merge styles and style variables for slots,
//        particularly if createComponent is modified not to generate classNames for slots
//
// const _resolveTokens = (props, ...allTokens) => Object.assign({}, ...allTokens.map(tokens =>
//    (typeof tokens === 'function') ? tokens(props) : tokens));
// const _resolveStyles = (props, tokens, ...allStyles) => concatStyleSets(...allStyles.map(styles =>
//    (typeof styles === 'function') ? styles(props, tokens) : styles));
//
// const createComponent = (options) => {
//   const component = componentProps => {
//     let { tokens, styles, ...props } = componentProps;
//
//     tokens = _resolveTokens(props, options.tokens, tokens);
//     styles = _resolveStyles(props, tokens, options.styles, styles);
//
//     const classNames = mergeStyleSets(styles);
//
//     return options.view({ ...props, classNames });
//   };
//   component.displayName = options.displayName;
//
//   return component;
// };
//////////////////////////////////////////////////////////////////////////////////////////////
