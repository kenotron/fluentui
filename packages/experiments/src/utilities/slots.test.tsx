import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { createElementWrapper, createFactory, getSlots, ISlot, ISlotProp, ISlotRenderFunction, ISlotDefinition, IUserProps } from './slots';

describe('createElementWrapper', () => {
  let reactCalls: number;

  beforeEach(() => {
    jest.spyOn(React, 'createElement').mockImplementation(() => {
      reactCalls += 1;
    });
    reactCalls = 0;
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('calls React.createElement by default', () => {
    createElementWrapper('span');

    expect(reactCalls).toEqual(1);
  });

  it('calls React.createElement without isSlot property present', () => {
    let factoryCalls = 0;
    const factoryComponent: ISlot<any> = () => {
      factoryCalls += 1;
      return <div />;
    };

    createElementWrapper(factoryComponent);

    expect(reactCalls).toEqual(1);
    expect(factoryCalls).toEqual(0);
  });

  it('does not call React.createElement with isSlot property present', () => {
    let factoryCalls = 0;
    const factoryComponent: ISlot<any> = () => {
      factoryCalls += 1;
      // prevent React.createElement from being called here by returning null
      return null as any;
    };
    factoryComponent.isSlot = true;

    createElementWrapper(factoryComponent);

    expect(reactCalls).toEqual(0);
    expect(factoryCalls).toEqual(1);
  });
});

describe('createFactory', () => {
  const componentProps = { component: 'componentValue', id: 'componentIdValue', children: ['Component Child 1', 'Component Child 2'] };
  const userProps = { user: 'userValue', id: 'userIdValue', children: ['User Child 1', 'User Child 2'] };
  const userPropString = 'userPropString';
  const defaultProp = 'shorthand-prop';
  const factoryOptions = { defaultProp };

  const TestComponent = (props: any) => {
    return <div {...props} />;
  };

  it(`passes componentProps without userProps`, () => {
    const component = mount(createFactory(TestComponent)(componentProps));
    expect(component.props()).toEqual(componentProps);
  });

  it(`passes userProp string as child`, () => {
    const component = mount(createFactory(TestComponent)(componentProps, userPropString));
    expect(component.props()).toEqual({ ...componentProps, children: userPropString });
  });

  it(`passes userProp integer as child`, () => {
    const component = mount(createFactory(TestComponent)(componentProps, 42));
    expect(component.props()).toEqual({ ...componentProps, children: 42 });
  });

  it(`passes userProp string as defaultProp`, () => {
    const component = mount(createFactory(TestComponent, factoryOptions)(componentProps, userPropString));
    expect(component.props()).toEqual({ ...componentProps, [defaultProp]: userPropString });
  });

  it(`passes userProp integer as defaultProp`, () => {
    const component = mount(createFactory(TestComponent, factoryOptions)(componentProps, 42));
    expect(component.props()).toEqual({ ...componentProps, [defaultProp]: 42 });
  });

  it('merges userProps over componentProps', () => {
    const component = mount(createFactory(TestComponent, factoryOptions)(componentProps, userProps));
    expect(component.props()).toEqual({ ...componentProps, ...userProps });
  });

  it('renders userProp integer as children', () => {
    const component = renderer.create(createFactory(TestComponent)(componentProps, 42));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders userProp string as children', () => {
    const component = renderer.create(createFactory(TestComponent)(componentProps, userPropString));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders userProp JSX with one prop', () => {
    const component = renderer.create(createFactory(TestComponent)(componentProps, <p id="I should be the only prop in the output" />));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders userProp function with one prop', () => {
    const component = renderer.create(
      createFactory(TestComponent)(componentProps, () => <p id="I should be the only prop in the output" />)
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`passes props and type arguments to userProp function`, done => {
    const userPropsFunction: ISlotRenderFunction<typeof userProps> = (props, type) => {
      expect(props).toEqual(componentProps);
      expect(type).toEqual(TestComponent);
      done();
      return <div {...props} />;
    };

    createFactory(TestComponent, factoryOptions)(componentProps, userPropsFunction);
  });
});

describe('getSlots', () => {
  interface ITestSlotComponent1Props {
    testSlot1Prop?: string;
  }
  interface ITestSlotComponent2Props {
    testSlot2Prop?: string;
  }
  interface ITestSlots {
    testSlot1?: ISlotProp<ITestSlotComponent1Props>;
    testSlot2?: ISlotProp<ITestSlotComponent2Props>;
  }

  interface ITestProps extends ITestSlots, IUserProps<ITestSlots> {}

  it(`creates slots and passes merged props to them`, done => {
    const testUserProps: ITestProps = {
      testSlot1: { testSlot1Prop: 'userProp1' },
      classNames: {
        testSlot1: 'testSlot1Classname',
        testSlot2: 'testSlot2Classname'
      }
    };

    const testSlot1Props: ITestSlotComponent1Props = {
      testSlot1Prop: 'slotProp1'
    };
    const testSlot2Props: ITestSlotComponent2Props = {
      testSlot2Prop: 'slotProp1'
    };

    const testSlotDefinition: ISlotDefinition<Required<ITestSlots>> = {
      testSlot1: props => {
        // User props should override slot props
        expect(props).toEqual({
          ...testUserProps.testSlot1,
          className: testUserProps.classNames.testSlot1
        });
        return null;
      },
      testSlot2: props => {
        // No user prop for slot in this case, so slot props should be present
        expect(props).toEqual({
          ...testSlot2Props,
          className: testUserProps.classNames.testSlot2
        });
        done();
        return null;
      }
    };

    const createdSlots = getSlots<ITestProps, ITestSlots>(testUserProps, testSlotDefinition);

    expect(createdSlots.testSlot1).toBeDefined();
    expect(createdSlots.testSlot1.isSlot).toBeTruthy();
    expect(createdSlots.testSlot2).toBeDefined();
    expect(createdSlots.testSlot2.isSlot).toBeTruthy();

    // Mount to trigger slot component render functions
    mount(createdSlots.testSlot1(testSlot1Props));
    mount(createdSlots.testSlot2(testSlot2Props));
  });
});
