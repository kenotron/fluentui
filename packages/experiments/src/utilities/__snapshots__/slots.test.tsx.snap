// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFactory renders userProp JSX with one prop 1`] = `
<p
  id="I should be the only prop in the output"
/>
`;

exports[`createFactory renders userProp function with one prop 1`] = `
<p
  id="I should be the only prop in the output"
/>
`;

exports[`createFactory renders userProp integer as children 1`] = `
<div
  component="componentValue"
  id="componentIdValue"
>
  42
</div>
`;

exports[`createFactory renders userProp string as children 1`] = `
<div
  component="componentValue"
  id="componentIdValue"
>
  userPropString
</div>
`;
